File: packages/agent/agent/tools/read_tool_events.ts
import { agentToolEvents } from "@clara-financas/db/schema/agent-event";
import { getDb } from "@clara-financas/db";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { and, desc, eq, type SQL } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { optionalText } from "../lib/schema";
import { requireTenantCaller } from "../lib/tenant";

/**
 * A telemetria ganha leitor — "o que deu errado nas últimas conversas?".
 *
 * `agent_tool_events` era escrita pelo hook e lida por ninguém: diagnosticar
 * dependia de `psql`. Com o leitor, a própria Clara responde "por que aquilo
 * falhou ontem?" citando a ferramenta, o código do erro e a duração — sem
 * expor nada além do que o resumo já protege por construção (chaves, ids e
 * contagens; nunca valor, descrição ou senha).
 *
 * Leitura pura, sem gate. A tabela é append-only por grant de banco.
 */
const LIMIT = 50;

export default defineTool({
  description:
    "Reads the tool execution log: which tool ran, status (ok/recuperavel/falha), error code, duration. Use for 'o que deu errado', 'por que falhou ontem'. Carries no financial values by construction — only tool names, ids and counts.",
  inputSchema: z.object({
    status: z
      .enum(["ok", "recuperavel", "falha"])
      .optional()
      .describe("Only events with this status. `falha` is the usual question."),
    toolName: optionalText().describe("Only events of one tool."),
    limit: z.number().int().min(1).max(LIMIT).optional(),
  }),
  async execute(input, ctx) {
    const { tenantId } = requireTenantCaller(ctx);
    const limit = input.limit ?? 20;

    const filters: SQL[] = [eq(agentToolEvents.tenantId, tenantId)];
    if (input.status !== undefined) filters.push(eq(agentToolEvents.status, input.status));
    if (input.toolName !== undefined) filters.push(eq(agentToolEvents.toolName, input.toolName));

    const rows = await forTenant(
      tenantId,
      async (tx) =>
        tx
          .select()
          .from(agentToolEvents)
          .where(and(...filters))
          .orderBy(desc(agentToolEvents.createdAt))
          .limit(limit + 1),
      getDb(),
    );

    const truncated = rows.length > limit;
    const page = truncated ? rows.slice(0, limit) : rows;

    return {
      count: page.length,
      ...(truncated
        ? {
            truncated: true as const,
            note: `Mostrando os ${limit} eventos mais recentes; há mais. Filtre por status ou toolName.`,
          }
        : {}),
      // Sucessos só são gravados com CLARA_TELEMETRY_ALL=1 — sem isso, a
      // ausência de eventos "ok" não significa que nada rodou.
      events: page.map((row) => ({
        toolName: row.toolName,
        status: row.status,
        errorCode: row.errorCode,
        errorMessage: row.errorMessage,
        durationMs: row.durationMs,
        inputSummary: row.inputSummary,
        at: row.createdAt.toISOString(),
      })),
    };
  },
});

File: services/agent/clara/tools/read_tool_events.py
"""A telemetria ganha leitor — porta de `agent/tools/read_tool_events.ts`.

`agent_tool_events` era escrita e lida por ninguém: diagnosticar dependia de
acesso direto ao banco. Leitura pura, sem gate — a tabela é append-only.
Carrega nada além do que o resumo já protege por construção: chaves, ids e
contagens, nunca valor, descrição ou senha.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from sqlalchemy import select
from sqlalchemy.orm import Session

from clara.db.models import AgentToolEvent

LIMIT = 50

EventStatus = Literal["ok", "recuperavel", "falha"]


@dataclass(frozen=True)
class ToolEventRow:
    tool_name: str
    status: str
    error_code: str | None
    error_message: str | None
    duration_ms: int | None
    input_summary: dict[str, object] | None
    at: str


@dataclass(frozen=True)
class ReadToolEventsResult:
    count: int
    events: list[ToolEventRow] = field(default_factory=list)
    truncated: bool = False
    note: str | None = None


def read_tool_events(
    session: Session,
    tenant_id: str,
    *,
    status: EventStatus | None = None,
    tool_name: str | None = None,
    limit: int = 20,
) -> ReadToolEventsResult:
    conditions = [AgentToolEvent.tenant_id == tenant_id]
    if status is not None:
        conditions.append(AgentToolEvent.status == status)
    if tool_name is not None:
        conditions.append(AgentToolEvent.tool_name == tool_name)

    rows = (
        session.execute(
            select(AgentToolEvent)
            .where(*conditions)
            .order_by(AgentToolEvent.created_at.desc())
            .limit(limit + 1)
        )
        .scalars()
        .all()
    )

    truncated = len(rows) > limit
    page = rows[:limit] if truncated else rows

    return ReadToolEventsResult(
        count=len(page),
        truncated=truncated,
        note=(
            f"Mostrando os {limit} eventos mais recentes; há mais. Filtre por status ou "
            "tool_name."
            if truncated
            else None
        ),
        events=[
            ToolEventRow(
                tool_name=row.tool_name,
                status=row.status,
                error_code=row.error_code,
                error_message=row.error_message,
                duration_ms=row.duration_ms,
                input_summary=row.input_summary,
                at=row.created_at.isoformat(),
            )
            for row in page
        ],
    )
